Add CLI for clipping cubes

mspelman07 opened the pull request:
Required for metoppv/mo-blue-team#516
Acceptance_test_PR: metoppv/improver_test_data#17

Adds in a CLI calling a pre-existing function in IMPROVER to clip data in a cube between an upper and lower bound. This is required to be able to clip cloud bases to 0m in the suite. This PR also includes some acceptance tests for the new CLI. 

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

brhooper commented on improver/cli/clip.py:
```diff
@@ -0,0 +1,60 @@
+def process(
+    cube: cli.inputcube, *, min_value: float = None, max_value: float = None,
+):
+    """Clip the data in the input cube to be below max_value and higher than min_value.
```
```suggestion
    """Clip the data in the input cube to be less than or equal to max_value and greater than or equal to min_value.
```
A minor point, but I think that the numpy `clip` function, and thus the IMPROVER `clip_cube_data` function [here](https://github.com/metoppv/improver/blob/master/improver/utilities/cube_manipulation.py), actually sets values higher than `max_value` to `max_value`, rather than enforcing them to be below `max_value`. Similarly, this is the case for values below `min_value`. 

The docstring could, alternatively, be updated to more closely align with the plugin docstring, referring to setting values above/below the max/min_value thresholds to max/min_value.

brhooper commented on improver/cli/clip.py:
```diff
@@ -0,0 +1,60 @@
+    Returns:
+        iris.cube.Cube:
+            A cube with the same metadata as the input cube but with the data clipped to be
+            higher than min_value and lower than max_value
```
```suggestion
            greater than or equal to min_value and less than or equal to max_value
```
See previous comment

brhooper reviewed (changes requested):
Thanks for this @mspelman07, I've just added minor comments about the docstrings. Other than that, the change looks fine and the new test seems to cover what needs testing.

mspelman07 commented:
I've updated the docsting as you mentioned to make it clearer. 

When I was running a suite I realised that the function `clip_cube_data ` only works on gridded data as it requires an x and y coordinate to slice over. This is done to reduce memory requirements but as spot data isn't as memory hungry np.clip can be called directly in the cli. I've added some accpetance tests for this as well.

brhooper commented on improver/cli/clip.py:
```diff
@@ -0,0 +1,68 @@
+
+    from improver.utilities.cube_manipulation import clip_cube_data, get_coord_names
+
+    if "spot_index" in get_coord_names(cube):
```
This will create more work for you (in terms of creating unit tests), but should this functionality be moved inside of the plugin, rather than here?

mspelman07 replied:
This part no longer exists following the changes I mention below

brhooper commented on improver/utilities/cube_manipulation.py:
```diff
@@ -550,7 +550,7 @@ def clip_cube_data(cube: Cube, minimum_value: float, maximum_value: float) -> Cu
 
     result = iris.cube.CubeList()
-    for cube_slice in cube.slices([cube.coord(axis="y"), cube.coord(axis="x")]):
+    for cube_slice in cube.slices(cube.coord(axis="x")):
```
I think the only other place this is used is in `‎improver/calibration/simple_bias_correction.py`. Is changing to only slice over the x axis likely to dramatically change performance in the CLI using this plugin? I don't know if we use this CLI in our IMPROVER suite, but I think the bias correction plugin and CLI were developed by BoM so I assume they might use it

mspelman07 replied:
I think you're right that this change would possibly effect the bias correct which I hadn't considered. Given this and the fact that the clip_cube_data seems to have been set up as a way to clip data using very little memory I have returned the functon to how it was previously. In the clip cli I now just call the np.clip for all cubes. I checked with the memory profiler and the maximum memory use when using a full cube of gridded data was 398.41MiB which seems to be acceptable.

brhooper reviewed (changes requested):
Thanks for those changes @mspelman07 and for noticing the need for further changes to allow clipping spot data. I've just added a couple of questions regarding these changes.

brhooper reviewed (dismissed):
Thanks for those changes @mspelman07. 

I'm not sure if there is an expectation that a CLI will call the corresponding IMPROVER plugin if such a thing exists. I understand that in this case we're trying to avoid the spatial slicing employed in the plugin, so I'm happy to approve this at this point in order to see if the 2nd reviewer has strong feelings about the situation.

A possible alternative to ignoring the plugin would be to add an additional input to the CLI and plugin along the lines of `spatial_slicing: Bool = True`, which then determines whether to call `np.clip` on the input array or to slice the input cube first. Having this default to true would avoid altering the behaviour of the simple bias correction [plugin](https://github.com/metoppv/improver/blob/master/improver/calibration/simple_bias_correction.py#L53).

gavinevans reviewed (changes requested):
Thanks @mspelman07 👍 

I've added one suggestion to standardise the naming of the files.

mspelman07 commented on improver_tests/acceptance/test_clip.py:
```diff
@@ -0,0 +1,57 @@
+def test_basic(tmp_path, min_value, max_value, cube_type):
+    """Test clip functionality with different combinations of min and max values """
+    kgo_dir = acc.kgo_root() / "clip" / cube_type
+    kgo_path = kgo_dir / f"kgo_{min_value,max_value}.nc"
```
I've updated the names as you suggested.
